buildSrc/src/main/java/org/elasticsearch/gradle/JdkDownloadPlugin.java

```diff
@@ -0,0 +1,187 @@
+                    }
+                    task.dependsOn(jdkConfig);
+
+                    // depend on the jdk directory "artifact" from the root project
+                    DependencyHandler dependencies = project.getDependencies();
+                    Map<String, Object> depConfig = new HashMap<>();
+                    depConfig.put("path", ":"); // root project
+                    depConfig.put("configuration", configName("localjdk", version, platform));
+                    dependencies.add(configurationName, dependencies.project(depConfig));
+
+                    // add jdkHome runtime property to the task
+                    Object jdkHomeGetter = new Object() {
+                        @Override
+                        public String toString() {
+                            return configurations.getByName(configurationName).getSingleFile().toString();
+                        }
+                    };
+                    extraProperties.set("jdkHome", jdkHomeGetter);
+                }
+            })
+        );
+    }
+
+    private static void setupRootJdkDownload(Project rootProject, String platform, String version) {
+        String extractTaskName = "extract" + capitalize(platform) + "Jdk" + version;
+        if (rootProject.getTasks().findByName(extractTaskName) != null) {
+            // already setup this version
+            return;
+        }
+
+        // decompose the bundled jdk version, broken into elements as: [feature, interim, update, build]
+        // Note the "patch" version is not yet handled here, as it has not yet been used by java.
+        Matcher jdkVersionMatcher = JDK_VERSION.matcher(version);
+        if (jdkVersionMatcher.matches() == false) {
+            throw new IllegalArgumentException("Malformed jdk version [" + version + "]");
+        }
+        String jdkVersion = jdkVersionMatcher.group(1) + (jdkVersionMatcher.group(2) != null ? (jdkVersionMatcher.group(2)) : "");
+        String jdkMajor = jdkVersionMatcher.group(1);
+        String jdkBuild = jdkVersionMatcher.group(3);
+
+        // add fake ivy repo for jdk url
+        String repoName = "jdk_repo_" + version;
+        if (rootProject.getRepositories().findByName(repoName) == null) {
+            rootProject.getRepositories().ivy(ivyRepo -> {
+                ivyRepo.setName(repoName);
+                ivyRepo.setUrl("https://download.java.net");
+                ivyRepo.patternLayout(layout -> {
+                    layout.artifact("java/GA/jdk" + jdkMajor + "/" + jdkBuild + "/GPL/openjdk-[revision]_[module]-x64_bin.[ext]");
+                });
+            });
+        }
+
+        // add the jdk as a "dependency"
+        final ConfigurationContainer configurations = rootProject.getConfigurations();
+        String remoteConfigName = configName("openjdk", version, platform);
+        String localConfigName = configName("localjdk", version, platform);
+        Configuration jdkConfig = configurations.findByName(remoteConfigName);
```

Comment:
I think we repeat this pattern more and more. I think all of these collections are `NamedDomainObjectCollection`s so I think we could add a `Boilerplate.fondOrCreate(NamedDomainObjectCollection collection, String name)` method to make it easier on the eyes.

Reply:
That may be useful, but I think this PR is big enough to warrant that as a separate discussion.